src/shared/non-empty-array.ts

```diff
@@ -0,0 +1,5 @@
+export type NonEmptyArray<T> = [T, ...T[]];
```

Comment:
I would prefer to avoid trying to encode non-emptiness of arrays in a type for the moment. As you can see in the discussion on https://stackoverflow.com/a/56006703, the simple implementation has behaviors that users will find surprising. Also this creates an extra decision point for new code that is written - should I use type or a regular array and handle possible emptiness at runtime?